File: GraphQLService/src/sensor/sensor.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { SensorEntity } from './sensor.entity';
import { CreateSensorInput } from './sensor.dto';


@Injectable()
export class SensorService {
    constructor(
        @InjectRepository(SensorEntity)
        private readonly repo: Repository<SensorEntity>,
    ) {}

    private getPagination(page: number, pageSize: number) {
    return {
        skip: (page - 1) * pageSize,
        take: pageSize,
        };
    }

    getLatest(deviceId: string) {
    return this.repo.findOne({
        where: { deviceId },
        order: { timestamp: 'DESC' },
    });
    }

    getByDevice(deviceId: string, page: number, pageSize: number) {
    return this.repo.find({
        where: { deviceId },
        ...this.getPagination(page, pageSize),
    });
    }

    getRange(from: number, to: number, page: number, pageSize: number) {
    return this.repo.find({
        where: {
        timestamp: (from as any, to as any),
        } as any,
        ...this.getPagination(page, pageSize),
    });
    }


    getTemperatureAbove(value: number, page: number, pageSize: number) {
    return this.repo.find({
        where: { temp: (value as any) },
        ...this.getPagination(page, pageSize),
    }).then(data => data.filter(x => x.temp > value));
    }

    getTemperatureBelow(value: number, page: number, pageSize: number) {
    return this.repo.find({
        where: { temp: (value as any) },
        ...this.getPagination(page, pageSize),
    }).then(data => data.filter(x => x.temp < value));
    }

    getHumidityAbove(value: number, page: number, pageSize: number) {
    return this.repo.find({
        ...this.getPagination(page, pageSize),
    }).then(data => data.filter(x => x.humidity > value));
    }

    getHumidityBelow(value: number, page: number, pageSize: number) {
    return this.repo.find({
        ...this.getPagination(page, pageSize),
    }).then(data => data.filter(x => x.humidity < value));
    }

    getCoAbove(value: number, page: number, pageSize: number) {
    return this.repo.find({
        ...this.getPagination(page, pageSize),
    }).then(data => data.filter(x => x.co > value));
    }

    getSmokeAbove(value: number, page: number, pageSize: number) {
    return this.repo.find({
        ...this.getPagination(page, pageSize),
    }).then(data => data.filter(x => x.smoke > value));
    }

    getLightOnly(deviceId: string, page: number, pageSize: number) {
    return this.repo.find({
        where: { deviceId },
        select: ['timestamp', 'light'],
        ...this.getPagination(page, pageSize),
    });
    }

    getTempOnly(deviceId: string, page: number, pageSize: number) {
    return this.repo.find({
        where: { deviceId },
        select: ['timestamp', 'temp'],
        ...this.getPagination(page, pageSize),
    });
    }

    getCoOnly(deviceId: string, page: number, pageSize: number) {
    return this.repo.find({
        where: { deviceId },
        select: ['timestamp', 'co'],
        ...this.getPagination(page, pageSize),
    });
    }

    getSmokeOnly(deviceId: string, page: number, pageSize: number) {
    return this.repo.find({
        where: { deviceId },
        select: ['timestamp', 'smoke'],
        ...this.getPagination(page, pageSize),
    });
    }

    getHumidityOnly(deviceId: string, page: number, pageSize: number) {
    return this.repo.find({
        where: { deviceId },
        select: ['timestamp', 'humidity'],
        ...this.getPagination(page, pageSize),
    });
    }

    getMonitoring(deviceId: string, page: number, pageSize: number) {
    return this.repo.find({
        where: { deviceId },
        select: [
        'timestamp',
        'temp',
        'humidity',
        'co',
        'smoke',
        'light',
        ],
        ...this.getPagination(page, pageSize),
    });
    }


    getCount() {
    return this.repo.count();
    }

    getDeviceCount(deviceId: string) {
    return this.repo.count({
        where: { deviceId },
    });
    }

    getStats() {
        return this.repo
            .createQueryBuilder('s')
            .select('MIN(s.temp)', 'minTemperature')
            .addSelect('MAX(s.temp)', 'maxTemperature')
            .addSelect('AVG(s.temp)', 'averageTemperature')

            .addSelect('MIN(s.co)', 'minCo')
            .addSelect('MAX(s.co)', 'maxCo')
            .addSelect('AVG(s.co)', 'averageCo')

            .addSelect('MIN(s.humidity)', 'minHumidity')
            .addSelect('MAX(s.humidity)', 'maxHumidity')
            .addSelect('AVG(s.humidity)', 'averageHumidity')

            .addSelect('MIN(s.smoke)', 'minSmoke')
            .addSelect('MAX(s.smoke)', 'maxSmoke')
            .addSelect('AVG(s.smoke)', 'averageSmoke')

            .addSelect('COUNT(*)', 'totalCount')
            .getRawOne();
    }

    async addSensorData(input: CreateSensorInput) {
    const sensor = this.repo.create({
        deviceId: input.deviceId,
        temp: input.temp,
        humidity: input.humidity,
        co: input.co,
        smoke: input.smoke,
        light: input.light,
        timestamp: input.timestamp,
    });

        return await this.repo.save(sensor);
    }

    async deleteSensorData(id: number) {
    const entity = await this.repo.findOne({
        where: { id },
    });

    if (!entity) {
        return false;
    }

    await this.repo.remove(entity);

    return true;
    }
}

File: gRPCService/Services/SensorService.cs
using Grpc.Core;
using gRPCService.Data;
using Microsoft.EntityFrameworkCore;

namespace gRPCService.Services
{
    public class SensorService : Sensor.SensorBase
    {
        private readonly AppDbContext _db;

        public SensorService(AppDbContext db)
        {
            _db = db;
        }

        public override async Task<SelectResponse> GetLatest(DeviceRequest request, ServerCallContext context)
        {
            var data = await _db.SensorData
                .AsNoTracking()
                .Where(x => x.DeviceId == request.DeviceId)
                .OrderByDescending(x => x.Timestamp)
                .FirstOrDefaultAsync();

            return new SelectResponse
            {
                Temp = data?.Temp ?? 0,
                Co = data?.Co ?? 0
            };
        }

        public override async Task<SensorResponse> GetByDevice(DevicePaginationRequest request, ServerCallContext context)
        {
            var data = await _db.SensorData
                .AsNoTracking()
                .Where(x => x.DeviceId == request.DeviceId)
                .Skip((request.Page - 1) * request.PageSize)
                .Take(request.PageSize)
                .ToListAsync();

            return Map(data);
        }

        public override async Task<SensorResponse> GetRange(RangeRequest request, ServerCallContext context)
        {
            var data = await _db.SensorData
                .AsNoTracking()
                .Where(x => x.Timestamp >= request.From &&
                            x.Timestamp <= request.To)
                .Skip((request.Page - 1) * request.PageSize)
                .Take(request.PageSize)
                .ToListAsync();

            return Map(data);
        }

        public override async Task<SensorResponse> GetTemperatureAbove(ValuePaginationRequest request, ServerCallContext context)
        {
            var data = await _db.SensorData
                .AsNoTracking()
                .Where(x => x.Temp > request.Value)
                .Skip((request.Page - 1) * request.PageSize)
                .Take(request.PageSize)
                .ToListAsync();

            return Map(data);
        }

        public override async Task<SensorResponse> GetTemperatureBelow(ValuePaginationRequest request, ServerCallContext context)
        {
            var data = await _db.SensorData
                .AsNoTracking()
                .Where(x => x.Temp < request.Value)
                .Skip((request.Page - 1) * request.PageSize)
                .Take(request.PageSize)
                .ToListAsync();

            return Map(data);
        }

        public override async Task<SensorResponse> GetHumidityAbove(ValuePaginationRequest request, ServerCallContext context)
        {
            var data = await _db.SensorData
                .AsNoTracking()
                .Where(x => x.Humidity > request.Value)
                .Skip((request.Page - 1) * request.PageSize)
                .Take(request.PageSize)
                .ToListAsync();

            return Map(data);
        }

        public override async Task<SensorResponse> GetHumidityBelow(ValuePaginationRequest request, ServerCallContext context)
        {
            var data = await _db.SensorData
                .AsNoTracking()
                .Where(x => x.Humidity < request.Value)
                .Skip((request.Page - 1) * request.PageSize)
                .Take(request.PageSize)
                .ToListAsync();

            return Map(data);
        }

        public override async Task<SensorResponse> GetCoAbove(ValuePaginationRequest request, ServerCallContext context)
        {
            var data = await _db.SensorData
                .AsNoTracking()
                .Where(x => x.Co > request.Value)
                .Skip((request.Page - 1) * request.PageSize)
                .Take(request.PageSize)
                .ToListAsync();

            return Map(data);
        }

        public override async Task<SensorResponse> GetSmokeAbove(ValuePaginationRequest request, ServerCallContext context)
        {
            var data = await _db.SensorData
                .AsNoTracking()
                .Where(x => x.Smoke > request.Value)
                .Skip((request.Page - 1) * request.PageSize)
                .Take(request.PageSize)
                .ToListAsync();

            return Map(data);
        }

        public override async Task<SensorResponse> GetByLightStatus(LightRequest request, ServerCallContext context)
        {
            var data = await _db.SensorData
                .AsNoTracking()
                .Where(x => x.Light == request.Status)
                .Skip((request.Page - 1) * request.PageSize)
                .Take(request.PageSize)
                .ToListAsync();

            return Map(data);
        }

        public override async Task<SensorResponse> GetLightOnly(DevicePaginationRequest request, ServerCallContext context)
        {
            var data = await _db.SensorData
                .AsNoTracking()
                .Where(x => x.DeviceId == request.DeviceId)
                .Select(x => new Models.SensorData
                {
                    Id = x.Id,
                    DeviceId = x.DeviceId,
                    Timestamp = x.Timestamp,
                    Light = x.Light
                })
                .Skip((request.Page - 1) * request.PageSize)
                .Take(request.PageSize)
                .ToListAsync();

            return Map(data);
        }

        public override async Task<SensorResponse> GetTempOnly(DevicePaginationRequest request, ServerCallContext context)
        {
            var data = await _db.SensorData
                .AsNoTracking()
                .Where(x => x.DeviceId == request.DeviceId)
                .Select(x => new Models.SensorData
                {
                    Id = x.Id,
                    DeviceId = x.DeviceId,
                    Timestamp = x.Timestamp,
                    Temp = x.Temp
                })
                .Skip((request.Page - 1) * request.PageSize)
                .Take(request.PageSize)
                .ToListAsync();

            return Map(data);
        }

        public override async Task<SensorResponse> GetCoOnly(DevicePaginationRequest request, ServerCallContext context)
        {
            var data = await _db.SensorData
                .AsNoTracking()
                .Where(x => x.DeviceId == request.DeviceId)
                .Select(x => new Models.SensorData
                {
                    Id = x.Id,
                    DeviceId = x.DeviceId,
                    Timestamp = x.Timestamp,
                    Co = x.Co
                })
                .Skip((request.Page - 1) * request.PageSize)
                .Take(request.PageSize)
                .ToListAsync();

            return Map(data);
        }

        public override async Task<SensorResponse> GetSmokeOnly(DevicePaginationRequest request, ServerCallContext context)
        {
            var data = await _db.SensorData
                .AsNoTracking()
                .Where(x => x.DeviceId == request.DeviceId)
                .Select(x => new Models.SensorData
                {
                    Id = x.Id,
                    DeviceId = x.DeviceId,
                    Timestamp = x.Timestamp,
                    Smoke = x.Smoke
                })
                .Skip((request.Page - 1) * request.PageSize)
                .Take(request.PageSize)
                .ToListAsync();

            return Map(data);
        }

        public override async Task<SensorResponse> GetHumidityOnly(DevicePaginationRequest request, ServerCallContext context)
        {
            var data = await _db.SensorData
                .AsNoTracking()
                .Where(x => x.DeviceId == request.DeviceId)
                .Select(x => new Models.SensorData
                {
                    Id = x.Id,
                    DeviceId = x.DeviceId,
                    Timestamp = x.Timestamp,
                    Humidity = x.Humidity
                })
                .Skip((request.Page - 1) * request.PageSize)
                .Take(request.PageSize)
                .ToListAsync();

            return Map(data);
        }

        public override async Task<SensorResponse> GetMonitoring(DevicePaginationRequest request, ServerCallContext context)
        {
            var data = await _db.SensorData
                .AsNoTracking()
                .Where(x => x.DeviceId == request.DeviceId)
                .Skip((request.Page - 1) * request.PageSize)
                .Take(request.PageSize)
                .ToListAsync();

            return Map(data);
        }

        public override async Task<CountResponse> GetDeviceCount(DeviceRequest request, ServerCallContext context)
        {
            var count = await _db.SensorData
                .CountAsync(x => x.DeviceId == request.DeviceId);

            return new CountResponse
            {
                Count = count
            };
        }

        public override async Task<StatsResponse> GetStats(Empty request, ServerCallContext context)
        {
            return new StatsResponse
            {
                MinTemperature = await _db.SensorData.MinAsync(x => x.Temp),
                MaxTemperature = await _db.SensorData.MaxAsync(x => x.Temp),
                AverageTemperature = await _db.SensorData.AverageAsync(x => x.Temp),

                MinCo = await _db.SensorData.MinAsync(x => x.Co),
                MaxCo = await _db.SensorData.MaxAsync(x => x.Co),
                AverageCo = await _db.SensorData.AverageAsync(x => x.Co),

                MinHumidity = await _db.SensorData.MinAsync(x => x.Humidity),
                MaxHumidity = await _db.SensorData.MaxAsync(x => x.Humidity),
                AverageHumidity = await _db.SensorData.AverageAsync(x => x.Humidity),

                MinSmoke = await _db.SensorData.MinAsync(x => x.Smoke),
                MaxSmoke = await _db.SensorData.MaxAsync(x => x.Smoke),
                AverageSmoke = await _db.SensorData.AverageAsync(x => x.Smoke),

                TotalCount = await _db.SensorData.CountAsync()
            };
        }

        private SensorResponse Map(List<Models.SensorData> data)
        {
            var response = new SensorResponse();

            response.Data.AddRange(data.Select(x => new SensorData
            {
                Id = x.Id,
                DeviceId = x.DeviceId,
                Temp = x.Temp,
                Humidity = x.Humidity,
                Co = x.Co,
                Smoke = x.Smoke,
                Light = x.Light,
                Timestamp = x.Timestamp
            }));

            return response;
        }

        public override async Task<SensorData> AddSensorData(CreateSensorRequest request, ServerCallContext context)
        {
            var entity = new Models.SensorData
            {
                DeviceId = request.DeviceId,
                Temp = request.Temp,
                Humidity = request.Humidity,
                Co = request.Co,
                Smoke = request.Smoke,
                Light = request.Light,
                Timestamp = request.Timestamp
            };

            _db.SensorData.Add(entity);

            await _db.SaveChangesAsync();

            return new SensorData
            {
                Id = entity.Id,
                DeviceId = entity.DeviceId,
                Temp = entity.Temp,
                Humidity = entity.Humidity,
                Co = entity.Co,
                Smoke = entity.Smoke,
                Light = entity.Light,
                Timestamp = entity.Timestamp
            };
        }

        public override async Task<DeleteResponse> DeleteSensorData(DeleteRequest request, ServerCallContext context)
        {
            var entity = await _db.SensorData
                .FirstOrDefaultAsync(x => x.Id == request.Id);

            if (entity == null)
            {
                return new DeleteResponse
                {
                    Success = false,
                    Message = "Sensor data not found"
                };
            }
            _db.SensorData.Remove(entity);

            await _db.SaveChangesAsync();

            return new DeleteResponse
            {
                Success = true,
                Message = "Deleted successfully"
            };
        }
    }
}